src/metabase-report-notifier/src/main/java/com/champtitles/metabasereportexecutor/notifier/EmailSender.java

```diff
@@ -0,0 +1,216 @@
+import org.slf4j.LoggerFactory;
+
+import javax.activation.DataHandler;
+import javax.activation.DataSource;
+import javax.mail.*;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeBodyPart;
+import javax.mail.internet.MimeMessage;
+import javax.mail.internet.MimeMultipart;
+import javax.mail.util.ByteArrayDataSource;
+import java.util.Properties;
+
+public class EmailSender {
+
+    private static final Logger logger = LoggerFactory.getLogger(EmailSender.class.getName());
+    private final Session session;
+    private final String fromAddress;
+    private final EmailTransport emailTransport;
+
+    /**
+     * @param smtpHost     host of SMTP server
+     * @param smtpPort     port of SMTP server
+     * @param smtpUser     username to log in to SMTP server
+     * @param smtpPassword password to log in to SMTP server
+     * @param fromAddress  email address to use as the sender
+     */
+    public EmailSender(String smtpHost, String smtpPort, String smtpUser, String smtpPassword, String fromAddress) {
+        this(smtpHost, smtpPort, smtpUser, smtpPassword, fromAddress, new EmailTransport());
+    }
+
+    EmailSender(String smtpHost, String smtpPort, String smtpUser, String smtpPassword, String fromAddress, EmailTransport emailTransport) {
+        Properties smtpProperties = createSmtpProperties(smtpHost, smtpPort, smtpUser, smtpPassword);
+        Authenticator smtpAuthenticator = createSmtpAuthenticator(smtpUser, smtpPassword);
+        session = Session.getInstance(smtpProperties, smtpAuthenticator);
+        this.fromAddress = fromAddress;
+        this.emailTransport = emailTransport;
+    }
+
+    /**
+     * Create an email with a xlsx attachment and send it
+     *
+     * @param subject    subject line of the email
+     * @param recipients list of recipient addresses
+     * @param fileName   name of the xlsx file attachment
+     * @param xlsxData   contents of the xlsx file attachment
+     */
+    public void sendEmail(String subject, String[] recipients, String fileName, byte[] xlsxData) {
+        logger.info("creating email message");
+        Message message = new MimeMessage(session);
+        setFromAddress(message);
+        setRecipients(message, recipients);
+        setSubject(message, subject);
+        setAttachment(message, fileName, xlsxData);
+        send(message);
+    }
+
+    /**
+     * Sends the email message using the transport
+     *
+     * @param message email message to send
+     */
+    private void send(Message message) {
+        try {
+            logger.info("sending email");
+            emailTransport.send(message);
+
+        } catch (MessagingException e) {
+            logger.error("failed to send email message");
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Set the FROM field in the email message
+     *
+     * @param message email message
+     */
+    private void setFromAddress(Message message) {
```

Comment:
Assuming these were broken out so that specific exceptions could be caught/logged?